x-pack/plugins/security_solution/public/detections/components/rules/step_define_rule/index.tsx

```diff
@@ -682,24 +678,33 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
     [setOptionsSelected]
   );
 
+  const assertFieldsExists = (obj: unknown): obj is DataViewFieldMap =>
+    !isEmpty(obj) && obj != null;
+
   const optionsData = useMemo(
     () =>
-      isEmpty(indexPattern.fields)
-        ? {
+      assertFieldsExists(indexPattern?.fields)
+        ? Object.values(indexPattern?.fields ?? {}).reduce(
+            (acc, field) => {
+              if (field?.esTypes?.includes('keyword')) {
+                return { ...acc, keywordFields: [...acc.keywordFields, { label: field.name }] };
+              } else if (field.type === 'date') {
+                return { ...acc, dateFields: [...acc.dateFields, { label: field.name }] };
+              } else if (field.type !== 'date') {
+                return { ...acc, nonDateFields: [...acc.nonDateFields, { label: field.name }] };
+              }
```

Comment:
```suggestion
             if (field?.esTypes?.includes('keyword')) {
              acc.keywordFields.push({ label: field.name });
            } else if (field.type === 'date') {
              acc.dateFields.push({ label: field.name });
            } else if (field.type !== 'date') {
              acc.nonDateFields.push({ label: field.name });
            }
```

number of fields returned from index can be very significant: just default Security indices return around 8k.
Having traversing them and then recreating array almost of the same size on each iteration can be very resource consuming.

I did some measurement and indeed it may cause issues

<img width="2554" alt="Screenshot 2023-06-23 at 12 20 24" src="https://github.com/elastic/kibana/assets/92328789/c686b6a0-ea59-4c3f-b2f4-04fa06748e3e">

Here, 7k fields took around 500ms to process.
With approach I propose, it takes ~ 4ms
<img width="2557" alt="Screenshot 2023-06-23 at 12 30 29" src="https://github.com/elastic/kibana/assets/92328789/f0cd830c-11dc-4447-8b94-24ed663bc637">

Reply:
awesome!!! Looks great thanks for this!